simplify setting op name for DifferentiabilityInfo instances

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63765

`dataclasses.replace()` allows us to replace only the fields we care
about without needing to explicitly copy over the remaining
fields. The former approach could be brittle if any fields have
default values or if we don't match up the names correctly.

Differential Revision: [D30489544](https://our.internmc.facebook.com/intern/diff/D30489544)

Commit:
simplify setting op name for DifferentiabilityInfo instances

`dataclasses.replace()` allows us to replace only the fields we care
about without needing to explicitly copy over the remaining
fields. The former approach could be brittle if any fields have
default values or if we don't match up the names correctly.

[ghstack-poisoned]

NicolasHug commented:
Hm 🤔

![image](https://user-images.githubusercontent.com/1190450/130466900-597b5d3d-2141-4adb-9e35-e723f481315d.png)


@dagitses did you push this from the AWS cluster? There has been multiple similar issues in the past: https://fb.workplace.com/groups/172776378014042

dagitses commented:
> Hm 🤔
> 
> ![image](https://user-images.githubusercontent.com/1190450/130466900-597b5d3d-2141-4adb-9e35-e723f481315d.png)
> 
> @dagitses did you push this from the AWS cluster? We might have an ssh key conflict, there has been multiple similar issues in the past: https://fb.workplace.com/groups/172776378014042

i did, and i have been seeing your name pop up on my PRs. I just assumed you were a fan of my work.

What do you think is going on?

NicolasHug commented:
There has been previous occurrences:

https://fb.workplace.com/groups/172776378014042/permalink/257820246176321/ and https://fb.workplace.com/groups/172776378014042/permalink/177793380845675

The issue is tracked in https://www.internalfb.com/tasks/?t=97042068 and the solution seems to be for you to re-generate your public ssh keys

dagitses commented:
> There has been previous occurrences:
> 
> https://fb.workplace.com/groups/172776378014042/permalink/257820246176321/ and https://fb.workplace.com/groups/172776378014042/permalink/177793380845675
> 
> The issue is tracked in https://www.internalfb.com/tasks/?t=97042068 and the solution seems to be for you to re-generate your public ssh keys

funny. ok, i think i fixed it. I did already generate keys, so I just removed ~/.ssh/id_rsa* and it seems to have done the trick.

albanD reviewed (approved):
Nice find!
There are a couple places where we do these that we should update as well!

dagitses commented:
> Nice find!
> There are a couple places where we do these that we should update as well!

i'll keep an eye out and replace them as i see them. I wonder if there's a way to use a backreference "regular" expression to find them, e.g. something like `(\w+)=\w+\.\1`

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30489544).

dagitses commented:
> > Nice find!
> > There are a couple places where we do these that we should update as well!
> 
> i'll keep an eye out and replace them as i see them. I wonder if there's a way to use a backreference "regular" expression to find them, e.g. something like `(\w+)=\w+\.\1`

`git grep --extended-regexp '(\w+)=\w+\.\1' tools/` shows some possible candidates.

dagitses commented:
> > > Nice find!
> > > There are a couple places where we do these that we should update as well!
> > 
> > 
> > i'll keep an eye out and replace them as i see them. I wonder if there's a way to use a backreference "regular" expression to find them, e.g. something like `(\w+)=\w+\.\1`
> 
> `git grep --extended-regexp '(\w+)=\w+\.\1' tools/` shows some possible candidates.

I came across a comment from Ed about `dataclasses.replace` in https://github.com/pytorch/pytorch/blob/master/tools/codegen/model.py#L1232:

> # dataclasses.replace could be used here, but it is less
> # type safe so for now I've opted to type everything out

I followed up and it fails to type check field assignments (e.g. it won't care about `dataclasses.replace(obj, some_str=7)`). That is unfortunate, but something that *can* improve in the future.

Which risks are you more concerned about?
1) forgetting to copy/clear a field
2) passing the wrong value to a field

Both types of breakages can occur. 1) will always fail if a field is missing that has no default value. This is typically the case in our code. That is a good thing because it is a very obvious failure that is easy to correct.

In the rare case that we forget a field with a default value, we may have a very subtle failure. This is unlikely in our codebase as I understand it.

Passing the wrong value to a field will likely cause a very obvious failure (if the type is not the same), but generally won't be immediate and will manifest when the field is used.

I don't think there's a strong robustness difference between the two cases, so I think we can go either way. And it may be that some cases are better for replace (e.g. we only want to clear a field) but others are better for naming all the fields (we have a lot of fields that are getting new values).

Do you want me to go forward with this change or withdraw it?

ezyang commented:
A reasonable compromise is to explicitly define `replace_op` on DifferentiabilityInfo; so that the method itself is strongly typed but the internal `replace` call is not. This minimizes the TCB so you only need to audit replace_op.

That being said, use of `dataclasses.replace` with this many arguments is somewhat of an antipattern, since it's slow (you are reconstructing a giant object) and suggests you messed up something about your data model (although I'm not entirely sure what happened here.)

dagitses commented:
> A reasonable compromise is to explicitly define `replace_op` on DifferentiabilityInfo; so that the method itself is strongly typed but the internal `replace` call is not. This minimizes the TCB so you only need to audit replace_op.

You have the exact same number of lines to audit, but it's easier since `self` and `op` will definitely be local and typed and it will be easier to see what the expected type of `self.op` is. Is that the argument?

> 
> That being said, use of `dataclasses.replace` with this many arguments is somewhat of an antipattern, since it's slow (you are reconstructing a giant object) and suggests you messed up something about your data model (although I'm not entirely sure what happened here.)

isn't it slow based on the number of fields of the object, not the number of fields you're modifying?

ezyang commented:
> You have the exact same number of lines to audit, but it's easier since self and op will definitely be local and typed and it will be easier to see what the expected type of self.op is. Is that the argument?

In this PR yes, but the moment there's another replace site then you also reduce the amount of auditing necessary

> isn't it slow based on the number of fields of the object, not the number of fields you're modifying?

Yes I misspoke, number of fields.

albanD commented:
> suggests you messed up something about your data model

In this case, it is more that we construct the whole object in phases. Filling in the fields one after the other.
Is that something we shouldn't be doing?

dagitses commented:
> Yeah, it's better not to, because it makes it less clear if you have a partially constructed object or fully constructed object in the code. Sometimes this is the only way to structure it but if there's a way to arrange for this to not be the deal (or have multiple differing types--the really heavyweight approach) that's better. I'm not familiar with the particularities here so I can't give much more detailed advice.

IMO, we should not add a method to the class for functionality that we already think is an anti-pattern and that is only used once. We also have the YAGNI principle here: we really aren't going to use this.

I looked into what it would take to refactor this to not need an intermediate partially constructed state: it is doable, but non-trivial.

On the bright side: to determine the op fields we only need the `name` and whether or not `args_with_derivatives` is non-empty.

On the very large negative side: we need this info for each entry in derivatives.yaml before we can proceed. This means we must do a minimum of two passes: we need to minimally first group all entries by name and identify which have args with derivatives.

Note that we already do two passes, it's just that the secondary pass is using partially constructed `DifferentiabilityInfo` instances.

One thing to consider is that *only* 24/576 of the entries do not have args with derivatives, so if we can filter these out cheaply then the problem becomes much easier. Are these functions considered nondifferentiable? If that's the case, what do you guys think about a property in native functions that indicates this instead, and then not requiring an entry in derivatives.yaml?

ezyang commented:
> If that's the case, what do you guys think about a property in native functions that indicates this instead, and then not requiring an entry in derivatives.yaml?

Something that we have been interested in, in the past, is merging `native_functions.yaml` and `derivatives.yaml`, to solve some "nontrivial coordination" problems where to understand whether or not a derivative was necessary or not required analysis of the two files separately. Previously, there was a stronger phase separation between stock ATen and autograd which merited the separation but this is less true now.

dagitses commented:
> > If that's the case, what do you guys think about a property in native functions that indicates this instead, and then not requiring an entry in derivatives.yaml?
> 
> Something that we have been interested in, in the past, is merging `native_functions.yaml` and `derivatives.yaml`, to solve some "nontrivial coordination" problems where to understand whether or not a derivative was necessary or not required analysis of the two files separately. Previously, there was a stronger phase separation between stock ATen and autograd which merited the separation but this is less true now.

here's an alternative implementation that breaks byte-for-byte compatibility but generates the names in a single forward pass. I think this is a superior solution if we're able to get away with it. #64261.

dagitses commented:
closing this as we have merged what we believe to be a superior solution. I'll audit other places for potential dataclasses.replace code smells.
